src/mlpack/methods/ann/loss_functions/poisson_nll_loss_impl.hpp

```diff
@@ -0,0 +1,96 @@
+/**
+ * @file poisson_nll_loss_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Poisson Negative Log Likelihood class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_POISSON_NLL_LOSS_IMPL_HPP
+
+
+// In case it hasn't yet been included.
+#include "poisson_nll_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+PoissonNLLLoss<InputDataType, OutputDataType>::PoissonNLLLoss(
+    const bool logInput,
+    const bool full,
+    const double eps,
+    const bool reduction):
+    logInput(logInput),
+    full(full),
+    eps(eps),
+    reduction(reduction)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+double PoissonNLLLoss<InputDataType, OutputDataType>::Forward(
+    const InputType& input,
+    TargetType& target)
+{
+  InputType loss;
+  loss.set_size(size(input));
+
+  if (logInput)
+    loss = arma::exp(input) - target % input;
+  else
+    loss = input - target % arma::log(input + eps);
+
+  if (full)
+  {
+    loss.elem(arma::find(target > 1)) += target % arma::log(target) - target
+        + arma::log(2 * M_PI * target) / 2;
+  }
+
+  if (reduction) return arma::accu(loss) / input.n_elem;
+  return arma::accu(loss);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType, typename OutputType>
+void PoissonNLLLoss<InputDataType, OutputDataType>::Backward(
+    const InputType& input,
+    const TargetType& target,
+    OutputType& output)
+{
+  output.set_size(size(input));
+  if (logInput) output = (arma::exp(input) - target);
+  else output = (1 - target / (input + eps));
```

Comment:
Same as above. I would prefer the statements in different lines. Most of the code-base follows that.